implementation/src/main/java/io/smallrye/config/SmallRyeConfig.java

```diff
@@ -228,40 +230,13 @@ private <T, C extends Collection<T>> C getIndexedValues(List<String> indexedProp
     }
 
     public List<String> getIndexedProperties(final String property) {
-        List<String> indexedProperties = new ArrayList<>();
-        for (String propertyName : this.getPropertyNames()) {
-            if (propertyName.startsWith(property) && propertyName.length() > property.length()) {
-                int indexStart = property.length();
-                if (propertyName.charAt(indexStart) == '[') {
-                    int indexEnd = propertyName.indexOf(']', indexStart);
-                    if (indexEnd != -1 && propertyName.charAt(propertyName.length() - 1) != '.'
-                            && StringUtil.isNumeric(propertyName, indexStart + 1, indexEnd)) {
-                        indexedProperties.add(propertyName);
-                    }
-                }
-            }
-        }
-        Collections.sort(indexedProperties);
-        return indexedProperties;
+        Map<Integer, String> indexedProperties = configSources.getPropertyNames().indexed().get(property);
+        return indexedProperties == null ? Collections.emptyList() : indexedProperties.values().stream().toList();
```

Comment:
With autoboxing, I doubt that a map is going to be better than a sorted list with all of the redundant object creation. Also, I think your method of converting to a stream is suspect. Are you *certain* this will give you your items in *index* order? Your test shows values in lexicographical order so it might be working accidentally.

Maybe the original problem is the actual sort though; if you're sorting numbers lexicographically, it's likely going to be a problem. It should not be too hard to define a numerically-sorting comparator that is efficient (find first non-zero character in each, compare lengths, then compare `Character.digit(x)` for each digit from most significant to least).

Reply:
The underlying map is backed by a `TreeMap`, and the index is being extracted, converted, and used as a key in the map (basically to discard invalid indexes).

The indexed structure is `Map<String, Map<Integer, String>>`, where the first key is the unindexed name, and then the `TreeMap` contains the indexes, with the target property. This was to avoid having to reconstruct the key.